ecscollector/collector.go

```diff
@@ -0,0 +1,216 @@
+	labels,
+	"cpu",
+)
+
+var networkLabels = append(
+	labels,
+	"device",
+)
+
+// NewCollector returns a new Collector that queries ECS metadata server
+// for ECS task and container metrics.
+func NewCollector(client *ecsmetadata.Client) prometheus.Collector {
+	return &collector{client: client}
+}
+
+type collector struct {
+	client *ecsmetadata.Client
+}
+
+func (c *collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- cpuTotalDesc
+	ch <- memUsageDesc
+	ch <- memMaxUsageDesc
+	ch <- memLimitDesc
+	ch <- networkRxBytesDesc
+	ch <- networkRxPacketsDesc
+	ch <- networkRxDroppedDesc
+	ch <- networkRxErrorsDesc
+	ch <- networkTxBytesDesc
+	ch <- networkTxPacketsDesc
+	ch <- networkTxDroppedDesc
+	ch <- networkTxErrorsDesc
+}
+
+func (c *collector) Collect(ch chan<- prometheus.Metric) {
+	ctx := context.Background()
+	metadata, err := c.client.RetrieveTaskMetadata(ctx)
+	if err != nil {
+		log.Printf("Failed to retrieve metadata: %v", err)
+		return
+	}
+	stats, err := c.client.RetrieveTaskStats(ctx)
+	if err != nil {
+		log.Printf("Failed to retrieve container stats: %v", err)
+		return
+	}
+	for _, container := range metadata.Containers {
+		s := stats[container.DockerID]
+		if s == nil {
+			log.Printf("Couldn't find container with ID %q in stats", container.DockerID)
+			continue
+		}
+
+		labelVals := []string{
+			container.Name,
+		}
+
+		for i, cpuUsage := range s.CPUStats.CPUUsage.PerCPUUsage {
+			cpu := fmt.Sprintf("%d", i)
+			ch <- prometheus.MustNewConstMetric(
+				cpuTotalDesc,
+				prometheus.CounterValue,
+				cpuJiffiesToSeconds(cpuUsage),
+				append(labelVals, cpu)...,
+			)
+		}
+
+		for desc, value := range map[*prometheus.Desc]float64{
+			memUsageDesc:    s.MemoryStats.Usage,
+			memMaxUsageDesc: s.MemoryStats.MaxUsage,
+			memLimitDesc:    s.MemoryStats.Limit,
+		} {
+			ch <- prometheus.MustNewConstMetric(
+				desc,
+				prometheus.GaugeValue,
+				value,
+				labelVals...,
+			)
+		}
+
+		// Network metrics per inteface.
+		for iface, netStats := range s.Networks {
+			networkLabelVals := append(labelVals, iface)
+
+			for desc, value := range map[*prometheus.Desc]float64{
+				networkRxBytesDesc:   netStats.RxBytes,
+				networkRxPacketsDesc: netStats.RxPackets,
+				networkRxDroppedDesc: netStats.RxDropped,
+				networkRxErrorsDesc:  netStats.RxErrors,
+				networkTxBytesDesc:   netStats.TxBytes,
+				networkTxPacketsDesc: netStats.TxPackets,
+				networkTxDroppedDesc: netStats.TxDropped,
+				networkTxErrorsDesc:  netStats.TxErrors,
+			} {
+				ch <- prometheus.MustNewConstMetric(
+					desc,
+					prometheus.CounterValue,
+					value,
+					networkLabelVals...,
+				)
+			}
+		}
+	}
+}
+
+// cpuJiffiesToSeconds converts CPU metrics
+// in jiffies to seconds.
+func cpuJiffiesToSeconds(j float64) float64 {
+	return j / float64(clockTick)
+}
+
+var clockTick int64 = 100 // Clock ticks are platform dependent, read from system config.
+
+func init() {
+	tick, err := sysconf.Sysconf(sysconf.SC_CLK_TCK)
```

Comment:
Interesting, I haven't seen this package before. This would be useful for `prometheus/procfs`.